PowerEditor/src/WinControls/ProjectPanel/DirectoryWatcher.h

```diff
@@ -0,0 +1,154 @@
+
+#include "Common.h"
+#include <CommCtrl.h>
+
+#include <map>
+#include <set>
+
+#include "resource.h"
+#include "Directory.h"
+#include "mutex.h"
+
+#define DIRECTORYWATCHER_UPDATE			DIRECTORYWATCHER_USER
+#define DIRECTORYWATCHER_UPDATE_DONE	DIRECTORYWATCHER_USER + 1
+
+// this class monitors specific directories for changes in a separate thread.
+// If a change occurs, it sends a DIRECTORYWATCHER_UPDATE message to the owner's window, with a tree item handle in lparam.
+//
+// It works by polling all directories in a given update frequency.
+// If changes are detected, and DIRECTORYWATCHER_UPDATE have been sent during a cycle, at last a DIRECTORYWATCHER_UPDATE_DONE is sent.
+//
+// The main part of this class is a thread, which periodically polls an arbitrary number of directories for changes.
+// I don't really like polling at all, but I tried out the two other obvious attempts, both of which had severe drawbacks:
+//
+// FindFirstChangeNotification/FindNextChangeNotification: Did not work very good, because a max of ~60 directories per thread
+// can be handled (64 objects limit of WaitForMultipleObjects)
+// To use a FindFirstChangeNotification for a whole tree is also not an option, because I wanted to allow to add root directories as well,
+// and e.g. the continuing changes of C: are insane (think of temp dir, logs etc.) And besides, FindFirstChangeNotification does not tell you WHAT has changed *sigh*
+//
+// GetQueuedCompletionStatus/ReadDirectoryChangesW: Worked perfect in terms of being notified, BUT locked all open directories against deletion, which is definitely not what we want.
+//
+// The polling is at least done as efficient as possible.
+// The first step is to compare the last write time of the directory.
+// Only if this has changed, the directory is examined more in depth, if any files have been added or removed.
+// Only if this is the case, the message is sent.
+//
+// Adding and removing directories works in two steps internally.
+// First, the directory is inserted to a set of directories to be removed or added.
+// Second, these sets are evaluated and its values are added or removed to the set of watched directories.
+// This is done so to keep the locking time as short as possible; only the transfering of the directories is locked, not the directory watch iteration itself.
+
+class DirectoryWatcher
+{
+	struct InsertStruct {
+		HTREEITEM _hTreeItem;
+		generic_string _path;
+		std::vector<generic_string> _filters;
+
+		InsertStruct(generic_string path, HTREEITEM hTreeItem, const std::vector<generic_string>& filters)
+			: _path(path)
+			, _hTreeItem(hTreeItem)
+			, _filters(filters)
+		{}
+	};
+
+	HANDLE _hThread;
+	HANDLE _hRunningEvent, _hStopEvent, _hUpdateEvent;
+	bool _running;
+	HWND _hWnd;
+	bool _hideEmptyDirs;
+	DWORD _checkEmptyDirsEvery;
+	DWORD _checkEmptyDirsCount;
+
+
+	std::set<Directory*> _watchdirs;
+	std::map<HTREEITEM, Directory*> _dirItems;
+	std::map<Directory*, int> _dirItemReferenceCount;
```

Comment:
I have to confess that I am not really used to smart pointers.
But it seems to be a very good idea, because it also removes the necessity to manually keep track of the reference count, since it comes for free with `shared_ptr`.
Will change it this way.